library/core/src/pin.rs

```diff
@@ -1246,5 +1237,16 @@ pub macro pin($value:expr $(,)?) {
     //
     // See https://doc.rust-lang.org/1.58.1/reference/destructors.html#temporary-lifetime-extension
     // for more info.
-    $crate::pin::Pin::<&mut _> { pointer: &mut { $value } }
+    //
+    // # Avoiding coercion
+    //
+    // We need to explicitly constrain the pointer type to avoid (unsound) coercion. Otherwise, you
+    // could write code like
+    // ```rust
+    // let rc = Rc::new(PhantomPinned);
+    // let _pinned: Pin<&PhantomPinned> = pin!(rc.clone());
```

Comment:
```suggestion
    // { let _pinned: Pin<&PhantomPinned> = pin!(rc.clone()); }
```

(otherwise the `Rc` clone temporary would be around by the time of the `into_inner()` call and the `.unwrap()` would fail)